Avoid a dispatch() in IPC sending if we're already processing outgoing messages

smfr opened the pull request:
#### 604de1a10add685f880174d4c16784f369c9e5eb
<pre>
Avoid a dispatch() in IPC sending if we&apos;re already processing outgoing messages
<a href="https://bugs.webkit.org/show_bug.cgi?id=258427">https://bugs.webkit.org/show_bug.cgi?id=258427</a>
rdar://111192703

Reviewed by Ben Nham and Chris Dumez.

Connection::sendMessage() would do a dispatch to the connection queue on every message
send, but this dispatch just triggers sendOutgoingMessages() to send things in the m_outgoingMessages
queue. So we only need to dispatch when this queue is empty.

This reduces the number of calls to dispatch() when loading amazon.com by about 18%.

If we&apos;re in the middle of dispatching normal-QOS messages and we&apos;re asked to dispach a higher QOS message,
dispatch anyway. It&apos;s not clear if this has the intended effect of boosting the priority of the already-running
dispatch work.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::sendMessage):

Canonical link: <a href="https://commits.webkit.org/265480@main">https://commits.webkit.org/265480@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1e321d5ff3fcc4221204f0fece02d610b31d467

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10896 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11109 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11422 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12543 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10420 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13486 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11090 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13336 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11055 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11956 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9179 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12947 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9249 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17074 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10320 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9992 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13229 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10439 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8526 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9611 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2644 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13881 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10299 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham reviewed (approved):
Nice. I had looked into this before, but I was looking at `CFRunLoopWakeUp` call in `RunLoop::dispatch` instead.

Seems like `resumeSendSource` properly takes care of the case of re-running `sendOutgoingMessages` if we break out of the sending loop early.

cdumez reviewed (approved):
LGTM

smfr committed:
Avoid a dispatch() in IPC sending if we're already processing outgoing messages
https://bugs.webkit.org/show_bug.cgi?id=258427
rdar://111192703

Reviewed by Ben Nham and Chris Dumez.

Connection::sendMessage() would do a dispatch to the connection queue on every message
send, but this dispatch just triggers sendOutgoingMessages() to send things in the m_outgoingMessages
queue. So we only need to dispatch when this queue is empty.

This reduces the number of calls to dispatch() when loading amazon.com by about 18%.

If we're in the middle of dispatching normal-QOS messages and we're asked to dispach a higher QOS message,
dispatch anyway. It's not clear if this has the intended effect of boosting the priority of the already-running
dispatch work.

* Source/WebKit/Platform/IPC/Connection.cpp:
(IPC::Connection::sendMessage):

Canonical link: https://commits.webkit.org/265480@main
